### ISSUE TITLE
Is libcryptsetup that thread-unsafe?

### ISSUE THREAD

**Issue by DemiMarie** [stratis-storage/libcryptsetup-rs#126]
According to https://gitlab.com/cryptsetup/cryptsetup/-/issues/69 libcryptsetup is thread-safe provided that each `crypt_device` object is not shared between threads.  If this is false, we might consider a reimplementation in Rust.

**Comment by jbaublitz** [comment]
I'm waiting on a more formal response from the libcryptsetup maintainers, but when dealing with activating devices, libcryptsetup is not thread-safe nor are the libraries that it depends on to do so.

**Comment by DemiMarie** [comment]
Yikes!  I take it that refers to libdevmapper?

**Comment by jbaublitz** [comment]
Yes that's correct! Our library provides the `mutex` feature to avoid undefined behavior in multithreaded environments by guarding all entry points into libcryptsetup but will essentially cause the accesses to be sequential.

**Comment by DemiMarie** [comment]
> Yes that's correct! Our library provides the `mutex` feature to avoid undefined behavior in multithreaded environments by guarding all entry points into libcryptsetup but will essentially cause the accesses to be sequential.

Someone really needs to fix libcryptsetup and libdevmapper.

**Comment by jbaublitz** [comment]
For now, I think our `mutex` feature is the only way around this. Please do let us know if there's forward motion on this and I'm happy to add some `Send`/`Sync` implementations as appropriate!

**Comment by jbaublitz** [comment]
@DemiMarie I can see the argument for both. On one hand, if it is not the default, it could be a little bit of a foot gun for people expecting it to be safe when called from separate threads. On the other hand, for single threaded applications this introduces completely unnecessary overhead. Would it be acceptable if I add detection for when it's called from multiple threads without the mutex feature and error out with a message instructing users to enable to `mutex` feature to continue?

**Comment by DemiMarie** [comment]
> @DemiMarie I can see the argument for both. On one hand, if it is not the default, it could be a little bit of a foot gun for people expecting it to be safe when called from separate threads. On the other hand, for single threaded applications this introduces completely unnecessary overhead. Would it be acceptable if I add detection for when it's called from multiple threads without the mutex feature and error out with a message instructing users to enable to `mutex` feature to continue?

That would be fine, pending a fix in libdevmapper.  In the future, maybe the solution is to either fix libdevmapper or to reimplement it in Rust.

**Comment by jbaublitz** [comment]
I'll file a separate issue for single thread validation!
